src/ModelParser/PhpDocParser.php

```diff
@@ -49,24 +60,37 @@ private function parseProperties(\ReflectionClass $reflClass, RawClassMetadata $
             }
 
             $docComment = $reflProperty->getDocComment();
-            if (false !== $docComment && $property->getType() instanceof PropertyTypeUnknown) {
+            if (false !== $docComment) {
                 try {
-                    $type = $this->getPropertyTypeFromDocComment($docComment, $reflClass);
+                    $type = $this->getPropertyTypeFromDocComment($docComment, $reflProperty);
                 } catch (ParseException $e) {
                     throw ParseException::propertyTypeError((string) $classMetadata, (string) $property, $e);
                 }
-                if (null !== $type) {
+                if (null === $type) {
+                    continue;
+                }
+
+                if ($property->getType() instanceof PropertyTypeUnknown || \array_key_exists((string) $property, $parentPropertiesLookup)) {
                     $property->setType($type);
+                } else {
+                    try {
+                        $property->setType($property->getType()->merge($type));
+                    } catch (\UnexpectedValueException $e) {
+                        throw ParseException::propertyTypeConflict((string) $classMetadata, (string) $property, (string) $property->getType(), (string) $type, $e);
+                    }
                 }
+                $addedProperties[] = (string) $property;
             }
         }
+
+        return array_values(array_diff(array_unique(array_merge($parentProperties, $addedProperties)), $existingProperties));
     }
 
-    private function getPropertyTypeFromDocComment(string $docComment, \ReflectionClass $reflClass): ?PropertyType
+    private function getPropertyTypeFromDocComment(string $docComment, \ReflectionProperty $reflProperty): ?PropertyType
     {
         foreach (explode("\n", $docComment) as $line) {
             if (1 === preg_match('/@var ([^ ]+)/', $line, $matches)) {
-                return $this->typeParser->parseAnnotationType($matches[1], $reflClass);
+                return $this->typeParser->parseAnnotationType($matches[1], $reflProperty->getDeclaringClass());
```

Comment:
we should add a test for the case where the declaring class is in a different namespace than the class we parse, and the property type is from the same namespace as the declaring base class.